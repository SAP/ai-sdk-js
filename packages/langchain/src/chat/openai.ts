import { CallbackManagerForLLMRun } from '@langchain/core/callbacks/manager';
import { BaseChatModelParams } from '@langchain/core/language_models/chat_models';
import { AIMessage, BaseMessage, ChatMessage, ToolMessage } from '@langchain/core/messages';
import { ChatResult } from '@langchain/core/outputs';
import { StructuredTool } from '@langchain/core/tools';
import { ChatOpenAI, ChatOpenAICallOptions, OpenAIChatInput } from '@langchain/openai';
import { zodToJsonSchema } from 'zod-to-json-schema';
import {
  OpenAiChatAssistantMessage,
  OpenAiChatToolMessage,
  OpenAiClient,
  OpenAiChatCompletionFunction,
  OpenAiChatFunctionCall,
  OpenAiChatMessage,
  OpenAiChatCompletionTool,
  OpenAiChatCompletionOutput,
  OpenAiChatCompletionParameters,
  DeploymentIdConfiguration,
  OpenAiChatModel
} from '@sap-ai-sdk/gen-ai-hub';

/**
 * Input for Text generation for OpenAI GPT.
 */
export interface OpenAIChatModelInterface
  extends Omit<OpenAIChatInput, 'openAIApiKey' | 'streaming'>,
    Omit<OpenAiChatCompletionParameters, 'n' | 'stop' | 'messages' | 'temperature'>,
    BaseChatModelParams {
      /**
       * The deployment ID of the model.
       */
      deploymentId?: string;
      /**
       * The version of the model.
       */
      modelVersion?: string;
    }

/**
 * Input for Text generation for OpenAI GPT.
 */
export type OpenAIChatModelInput = Omit<OpenAIChatInput,
  'frequencyPenalty'
  | 'presencePenalty'
  | 'topP'
  | 'temperature'
  | 'stop'
  | 'n'
  | 'modelName'
  | 'model'
  | 'openAIApiKey'
  | 'streaming'> &
  Omit<OpenAiChatCompletionParameters, 'messages'> &
  BaseChatModelParams &
  DeploymentIdConfiguration &
  {
    /**
     * The name of the model.
     */
    modelName: OpenAiChatModel;
    /**
     * The version of the model.
     */
    modelVersion?: string;
    deploymentId?: string;
  };

/**
 * Chat Call options.
 */
interface OpenAIChatCallOptions
  extends Omit<ChatOpenAICallOptions, 'promptIndex' | 'functions' | 'function_call' | 'tools'> {
  functions?: OpenAiChatCompletionFunction[];
  function_call?: OpenAiChatFunctionCall;
  tools?: OpenAiChatCompletionTool[];
}

/**
 * OpenAI Language Model Wrapper to generate texts.
 */
export class OpenAIChat extends ChatOpenAI implements OpenAIChatModelInterface {
  declare CallOptions: OpenAIChatCallOptions;

  deploymentId?: string;
  modelVersion?: string;
  private btpOpenAIClient: OpenAiClient;

  constructor(fields: OpenAIChatModelInput) {
    const defaultValues = new ChatOpenAI();
    const n = fields.n ?? defaultValues.n;
    const stop = fields.stop ? Array.isArray(fields.stop) ? fields.stop : [fields.stop] : defaultValues.stop;
    const temperature = fields.temperature ?? defaultValues.temperature;
    const frequencyPenalty = fields.frequency_penalty ?? defaultValues.frequencyPenalty;
    const presencePenalty = fields.presence_penalty ?? defaultValues.presencePenalty;
    const topP = fields.top_p ?? defaultValues.topP;

    super({
      ...fields,
      n,
      stop,
      temperature,
      openAIApiKey: 'dummy',
      frequencyPenalty,
      presencePenalty,
      topP
    });

    this.deploymentId = fields.deploymentId;
    this.modelVersion = fields.modelVersion;

    // LLM client
    this.btpOpenAIClient = new OpenAiClient();
  }

  override get callKeys(): (keyof OpenAIChatCallOptions)[] {
    return [
      ...(super.callKeys as (keyof OpenAIChatCallOptions)[]),
      'options',
      'function_call',
      'functions',
      'tools',
      'tool_choice',
      'response_format',
      'seed',
    ];
  }

  override get lc_secrets(): { [key: string]: string } | undefined {
    // overrides default keys as not applicable in BTP
    return {};
  }

  override get lc_aliases(): Record<string, string> {
    // overrides default keys as not applicable in BTP
    return {};
  }

  override async _generate(
    messages: BaseMessage[],
    options: this['CallOptions'],
    runManager?: CallbackManagerForLLMRun,
  ): Promise<ChatResult> {
    function isStructuredToolArray(tools?: unknown[]): tools is StructuredTool[] {
      return tools !== undefined && tools.every((tool) => Array.isArray((tool as StructuredTool).lc_namespace));
    }
    const res = await this.caller.callWithOptions(
      {
        signal: options.signal,
      },
      () =>
        this.btpOpenAIClient.chatCompletion(
          {
            messages: messages.map(this.mapBaseMessageToOpenAIChatMessage.bind(this)),
            deployment_id: this.deploymentId,
            max_tokens: this.maxTokens === -1 ? undefined : this.maxTokens,
            temperature: this.temperature,
            top_p: this.topP,
            logit_bias: this.logitBias,
            n: this.n,
            stop: options?.stop ?? this.stop,
            presence_penalty: this.presencePenalty,
            frequency_penalty: this.frequencyPenalty,
            functions: isStructuredToolArray(options?.functions)
              ? options?.functions.map(this.mapToolToBTPOpenAIFunction.bind(this))
              : options?.functions,
            tools: isStructuredToolArray(options?.tools)
              ? options?.tools.map(this.mapToolToOpenAITool.bind(this))
              : options?.tools,
            tool_choice: options?.tool_choice,
            response_format: options?.response_format,
            seed: options?.seed,
            ...this.modelKwargs,
          },
          {
            modelName: this.modelName ?? this.model,
            deploymentId: this.deploymentId,
            modelVersion: this.modelVersion
          },
        ),
    );

    // currently BTP LLM Proxy for OpenAI doesn't support streaming
    await runManager?.handleLLMNewToken(
      typeof res.choices[0].message.content === 'string' ? res.choices[0].message.content : '',
    );

    return this.mapResponseToChatResult(res);
  }

  /**
   * Maps a LangChain {@link StructuredTool} to {@link OpenAiChatCompletionFunction}.
   * @param tool - Base class for Tools that accept input of any shape defined by a Zod schema.
   * @returns The OpenAI Chat Completion Function.
   */
  protected mapToolToBTPOpenAIFunction(tool: StructuredTool): OpenAiChatCompletionFunction {
    return {
      name: tool.name,
      description: tool.description,
      parameters: zodToJsonSchema(tool.schema),
    };
  }

  /**
   * Maps a LangChain {@link StructuredTool} to {@link OpenAiChatCompletionTool}.
   * @param tool - Base class for Tools that accept input of any shape defined by a Zod schema.
   * @returns The OpenAI Chat Completion Tool.
   */
  protected mapToolToOpenAITool(tool: StructuredTool): OpenAiChatCompletionTool {
    return {
      type: 'function',
      function: {
        name: tool.name,
        description: tool.description,
        parameters: zodToJsonSchema(tool.schema),
      },
    };
  }

  /**
   * Maps a {@link BaseMessage} to OpenAI's Message Role.
   * @param message - The message to map.
   * @returns The OpenAI Message Role.
   */
  protected mapBaseMessageToRole(message: BaseMessage): OpenAiChatMessage['role'] {
    switch (message._getType()) {
      case 'ai':
        return 'assistant';
      case 'human':
        return 'user';
      case 'system':
        return 'system';
      case 'function':
        return 'function';
      case 'tool':
        return 'tool';
      case 'generic':
        return (message as ChatMessage).role as OpenAiChatMessage['role'];
      default:
        throw new Error(`Unknown message type: ${message._getType()}`);
    }
  }

  /**
   * Maps {@link BaseMessage} to OpenAI Messages.
   * @param message - The message to map.
   * @returns The OpenAI Chat Message.
   */
  protected mapBaseMessageToOpenAIChatMessage(message: BaseMessage): OpenAiChatMessage {
    return {
      content: message.content,
      name: message.name,
      role: this.mapBaseMessageToRole(message),
      function_call: message.additional_kwargs.function_call,
      tool_calls: message.additional_kwargs.tool_calls,
      tool_call_id: (message as ToolMessage).tool_call_id,
    } as OpenAiChatMessage;
  }

  /**
   * Maps OpenAI messages to LangChain's {@link ChatResult}.
   * @param res - The OpenAI Chat Completion Output.
   * @returns The LangChain Chat Result.
   */
  protected mapResponseToChatResult(res: OpenAiChatCompletionOutput): ChatResult {
    return {
      generations: res.choices.map((c) => ({
          text: (c.message as OpenAiChatAssistantMessage).content || '',
          message: new AIMessage({
            content: (c.message as OpenAiChatAssistantMessage).content || '',
            additional_kwargs: {
              finish_reason: c.finish_reason,
              index: c.index,
              logprobs: c.logprobs,
              function_call: (c.message as OpenAiChatAssistantMessage).function_call, // add `function_call` parameter
              tool_calls: (c.message as OpenAiChatAssistantMessage).tool_calls,
              tool_call_id: (c.message as OpenAiChatToolMessage).tool_call_id,
            },
          }),
          generationInfo: {
            finish_reason: c.finish_reason,
            index: c.index,
            logprobs: c.logprobs,
            function_call: (c.message as OpenAiChatAssistantMessage).function_call, // add `function_call` parameter
            tool_calls: (c.message as OpenAiChatAssistantMessage).tool_calls,
          },
        })),
      llmOutput: {
        created: res.created,
        id: res.id,
        model: res.model,
        object: res.object,
        tokenUsage: {
          completionTokens: res.usage.completion_tokens,
          promptTokens: res.usage.prompt_tokens,
          totalTokens: res.usage.total_tokens,
        },
      },
    };
  }
}
