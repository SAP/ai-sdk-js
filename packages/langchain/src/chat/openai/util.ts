import {
  AIMessage,
  BaseMessage,
  ChatMessage,
  ToolMessage
} from '@langchain/core/messages';
import { ChatResult } from '@langchain/core/outputs';
import { StructuredTool } from '@langchain/core/tools';
import {
  OpenAiChatAssistantMessage,
  OpenAiChatCompletionChoice,
  OpenAiChatCompletionFunction,
  OpenAiChatCompletionOutput,
  OpenAiChatCompletionTool,
  OpenAiChatMessage,
  OpenAiChatToolMessage
} from '@sap-ai-sdk/gen-ai-hub';
import { zodToJsonSchema } from 'zod-to-json-schema';

/**
 * Maps a LangChain {@link StructuredTool} to {@link OpenAiChatCompletionFunction}.
 * @param tool - Base class for Tools that accept input of any shape defined by a Zod schema.
 * @returns The OpenAI Chat Completion Function.
 */
export function mapToolToOpenAIFunction(
  tool: StructuredTool
): OpenAiChatCompletionFunction {
  return {
    name: tool.name,
    description: tool.description,
    parameters: zodToJsonSchema(tool.schema)
  };
}

/**
 * Maps a LangChain {@link StructuredTool} to {@link OpenAiChatCompletionTool}.
 * @param tool - Base class for Tools that accept input of any shape defined by a Zod schema.
 * @returns The OpenAI Chat Completion Tool.
 */
export function mapToolToOpenAITool(
  tool: StructuredTool
): OpenAiChatCompletionTool {
  return {
    type: 'function',
    function: {
      name: tool.name,
      description: tool.description,
      parameters: zodToJsonSchema(tool.schema)
    }
  };
}

/**
 * Maps a {@link BaseMessage} to OpenAI's Message Role.
 * @param message - The message to map.
 * @returns The OpenAI Message Role.
 */
export function mapBaseMessageToRole(
  message: BaseMessage
): OpenAiChatMessage['role'] {
  switch (message._getType()) {
    case 'ai':
      return 'assistant';
    case 'human':
      return 'user';
    case 'system':
      return 'system';
    case 'function':
      return 'function';
    case 'tool':
      return 'tool';
    case 'generic':
      return (message as ChatMessage).role as OpenAiChatMessage['role'];
    default:
      throw new Error(`Unknown message type: ${message._getType()}`);
  }
}

/**
 * Maps OpenAI messages to LangChain's {@link ChatResult}.
 * @param res - The OpenAI Chat Completion Output.
 * @returns The LangChain Chat Result.
 */
export function mapResponseToChatResult(
  res: OpenAiChatCompletionOutput
): ChatResult {
  return {
    generations: res.choices.map((c: OpenAiChatCompletionChoice) => ({
      text: (c.message as OpenAiChatAssistantMessage).content || '',
      message: new AIMessage({
        content: (c.message as OpenAiChatAssistantMessage).content || '',
        additional_kwargs: {
          finish_reason: c.finish_reason,
          index: c.index,
          function_call: (c.message as OpenAiChatAssistantMessage)
            .function_call, // add `function_call` parameter
          tool_calls: (c.message as OpenAiChatAssistantMessage).tool_calls,
          tool_call_id: (c.message as OpenAiChatToolMessage).tool_call_id
        }
      }),
      generationInfo: {
        finish_reason: c.finish_reason,
        index: c.index,
        function_call: (c.message as OpenAiChatAssistantMessage).function_call, // add `function_call` parameter
        tool_calls: (c.message as OpenAiChatAssistantMessage).tool_calls
      }
    })),
    llmOutput: {
      created: res.created,
      id: res.id,
      model: res.model,
      object: res.object,
      tokenUsage: {
        completionTokens: res.usage.completion_tokens,
        promptTokens: res.usage.prompt_tokens,
        totalTokens: res.usage.total_tokens
      }
    }
  };
}

/**
 * Maps {@link BaseMessage} to OpenAI Messages.
 * @param message - The message to map.
 * @returns The OpenAI Chat Message.
 */
export function mapBaseMessageToOpenAIChatMessage(
  message: BaseMessage
): OpenAiChatMessage {
  return {
    content: message.content,
    name: message.name,
    role: mapBaseMessageToRole(message),
    function_call: message.additional_kwargs.function_call,
    tool_calls: message.additional_kwargs.tool_calls,
    tool_call_id: (message as ToolMessage).tool_call_id
  } as OpenAiChatMessage;
}
