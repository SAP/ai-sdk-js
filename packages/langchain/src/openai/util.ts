import { AIMessage, ToolMessage } from '@langchain/core/messages';
import { zodToJsonSchema } from 'zod-to-json-schema';
import type { BaseMessage } from '@langchain/core/messages';
import type { ChatResult } from '@langchain/core/outputs';
import type { StructuredTool } from '@langchain/core/tools';
import type {
  AzureOpenAiChatCompletionRequestMessageSystem,
  AzureOpenAiChatCompletionRequestMessageUser,
  AzureOpenAiChatCompletionRequestMessageAssistant,
  AzureOpenAiChatCompletionRequestMessageTool,
  AzureOpenAiChatCompletionRequestMessageFunction,
  AzureOpenAiChatCompletionTool,
  AzureOpenAiChatCompletionRequestMessage,
  AzureOpenAiCreateChatCompletionResponse,
  AzureOpenAiCreateChatCompletionRequest,
  AzureOpenAiChatCompletionFunctionParameters
} from '@sap-ai-sdk/foundation-models';
import type { AzureOpenAiChatClient } from './chat.js';
import type { AzureOpenAiChatCallOptions } from './types.js';

type ToolChoice =
  | 'none'
  | 'auto'
  | {
      /**
       * The type of the tool.
       */
      type: 'function';
      /**
       * Use to force the model to call a specific function.
       */
      function: {
        /**
         * The name of the function to call.
         */
        name: string;
      };
    };

type LangChainToolChoice = string | Record<string, any> | 'auto' | 'any';

/**
 * Maps a LangChain {@link StructuredTool} to {@link AzureOpenAiChatCompletionFunction}.
 * @param tool - Base class for tools that accept input of any shape defined by a Zod schema.
 * @returns The OpenAI chat completion function.
 */
function mapToolToOpenAiFunction(tool: StructuredTool): {
  description?: string;
  name: string;
  parameters: AzureOpenAiChatCompletionFunctionParameters;
} & Record<string, any> {
  return {
    name: tool.name,
    description: tool.description,
    parameters: zodToJsonSchema(tool.schema)
  };
}

/**
 * Maps a LangChain {@link StructuredTool} to {@link AzureOpenAiChatCompletionTool}.
 * @param tool - Base class for tools that accept input of any shape defined by a Zod schema.
 * @returns The OpenAI chat completion tool.
 */
function mapToolToOpenAiTool(
  tool: StructuredTool
): AzureOpenAiChatCompletionTool {
  return {
    type: 'function',
    function: mapToolToOpenAiFunction(tool)
  };
}

/**
 * Maps a {@link BaseMessage} to{@link AzureOpenAiChatMessage} message role.
 * @param message - The {@link BaseMessage} to map.
 * @returns The {@link AzureOpenAiChatMessage} message Role.
 */
function mapBaseMessageToRole(
  message: BaseMessage
): AzureOpenAiChatCompletionRequestMessage['role'] {
  const messageTypeToRoleMap = new Map<
    string,
    AzureOpenAiChatCompletionRequestMessage['role']
  >([
    ['human', 'user'],
    ['ai', 'assistant'],
    ['system', 'system'],
    ['function', 'function'],
    ['tool', 'tool']
  ]);

  const messageType = message._getType();
  const role = messageTypeToRoleMap.get(messageType);
  if (!role) {
    throw new Error(`Unsupported message type: ${messageType}`);
  }
  return role;
}

/**
 * Maps {@link AzureOpenAiCreateChatCompletionResponse} to LangChain's {@link ChatResult}.
 * @param completionResponse - The {@link AzureOpenAiCreateChatCompletionResponse} response.
 * @returns The LangChain {@link ChatResult}
 * @internal
 */
export function mapOutputToChatResult(
  completionResponse: AzureOpenAiCreateChatCompletionResponse
): ChatResult {
  return {
    generations: completionResponse.choices.map(
      (choice: (typeof completionResponse)['choices'][0]) => ({
        text: choice.message?.content || '',
        message: new AIMessage({
          content: choice.message?.content || '',
          additional_kwargs: {
            finish_reason: choice.finish_reason,
            index: choice.index,
            function_call: choice.message?.function_call,
            tool_calls: choice.message?.tool_calls,
            tool_call_id: ''
          }
        }),
        generationInfo: {
          finish_reason: choice.finish_reason,
          index: choice.index,
          function_call: choice.message?.function_call,
          tool_calls: choice.message?.tool_calls
        }
      })
    ),
    llmOutput: {
      created: completionResponse.created,
      id: completionResponse.id,
      model: completionResponse.model,
      object: completionResponse.object,
      tokenUsage: {
        completionTokens: completionResponse.usage?.completion_tokens || 0,
        promptTokens: completionResponse.usage?.prompt_tokens || 0,
        totalTokens: completionResponse.usage?.total_tokens || 0
      }
    }
  };
}

/**
 * Maps {@link BaseMessage} to {@link AzureOpenAiChatMessage}.
 * @param message - The message to map.
 * @returns The {@link AzureOpenAiChatMessage}.
 */
function mapBaseMessageToAzureOpenAiChatMessage(
  message: BaseMessage
): AzureOpenAiChatCompletionRequestMessage {
  return removeUndefinedProperties<AzureOpenAiChatCompletionRequestMessage>({
    name: message.name,
    ...mapRoleAndContent(message),
    function_call: message.additional_kwargs.function_call,
    tool_calls: message.additional_kwargs.tool_calls,
    tool_call_id: mapToolCallId(message)
  });
}

// The following types are used to match a role to its specific content, otherwise TypeScript would not be able to infer the content type.

type Role = 'system' | 'user' | 'assistant' | 'tool' | 'function';

type ContentType<T extends Role> = T extends 'system'
  ? AzureOpenAiChatCompletionRequestMessageSystem['content']
  : T extends 'user'
    ? AzureOpenAiChatCompletionRequestMessageUser['content']
    : T extends 'assistant'
      ? AzureOpenAiChatCompletionRequestMessageAssistant['content']
      : T extends 'tool'
        ? AzureOpenAiChatCompletionRequestMessageTool['content']
        : T extends 'function'
          ? AzureOpenAiChatCompletionRequestMessageFunction['content']
          : never;

type RoleAndContent = {
  [T in Role]: { role: T; content: ContentType<T> };
}[Role];

function mapRoleAndContent(baseMessage: BaseMessage): RoleAndContent {
  const role = mapBaseMessageToRole(baseMessage);
  return {
    role,
    content: baseMessage.content as ContentType<typeof role>
  } as RoleAndContent;
}

function isStructuredToolArray(tools?: unknown[]): tools is StructuredTool[] {
  return !!tools?.every(tool =>
    Array.isArray((tool as StructuredTool).lc_namespace)
  );
}

/**
 * Has to return an empty string to match one of the types of {@link AzureOpenAiChatCompletionRequestMessage}.
 * @internal
 */
function mapToolCallId(message: BaseMessage): string {
  return ToolMessage.isInstance(message) ? message.tool_call_id : '';
}

function mapToolChoice(
  toolChoice?: LangChainToolChoice
): ToolChoice | undefined {
  if (toolChoice === 'auto' || toolChoice === 'none') {
    return toolChoice;
  }

  if (typeof toolChoice === 'string') {
    return {
      type: 'function',
      function: { name: toolChoice }
    };
  }
}

/**
 * Maps LangChain's input interface to the AI SDK client's input interface
 * @param client The LangChain Azure OpenAI client
 * @param options The {@link AzureOpenAiChatCallOptions}
 * @param messages The messages to be send
 * @returns An AI SDK compatibile request
 * @internal
 */
export function mapLangchainToAiClient(
  client: AzureOpenAiChatClient,
  messages: BaseMessage[],
  options?: AzureOpenAiChatCallOptions & { promptIndex?: number }
): AzureOpenAiCreateChatCompletionRequest {
  return removeUndefinedProperties<AzureOpenAiCreateChatCompletionRequest>({
    messages: messages.map(mapBaseMessageToAzureOpenAiChatMessage),
    max_tokens: client.max_tokens === -1 ? undefined : client.max_tokens,
    presence_penalty: client.presence_penalty,
    frequency_penalty: client.frequency_penalty,
    temperature: client.temperature,
    top_p: client.top_p,
    logit_bias: client.logit_bias,
    user: client.user,
    data_sources: options?.data_sources,
    n: options?.n,
    response_format: options?.response_format,
    seed: options?.seed,
    logprobs: options?.logprobs,
    top_logprobs: options?.top_logprobs,
    function_call: options?.function_call,
    stop: options?.stop ?? client.stop,
    functions: isStructuredToolArray(options?.functions)
      ? options?.functions.map(mapToolToOpenAiFunction)
      : options?.functions,
    tools: isStructuredToolArray(options?.tools)
      ? options?.tools.map(mapToolToOpenAiTool)
      : options?.tools,
    tool_choice: mapToolChoice(options?.tool_choice)
  });
}

function removeUndefinedProperties<T extends object>(obj: T): T {
  const result = { ...obj };
  for (const key in result) {
    if (result[key as keyof T] === undefined) {
      delete result[key as keyof T];
    }
  }
  return result;
}
