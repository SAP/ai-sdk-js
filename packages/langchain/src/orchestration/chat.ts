import { BaseChatModel } from '@langchain/core/language_models/chat_models';
import { OrchestrationClient as OrchestrationClientBase } from '@sap-ai-sdk/orchestration';
import { AsyncCaller } from '@langchain/core/utils/async_caller';
import { resilience } from '@sap-cloud-sdk/resilience';
import {
  isTemplate,
  mapLangchainMessagesToOrchestrationMessages,
  mapOutputToChatResult
} from './util.js';
import type { CustomRequestConfig } from '@sap-ai-sdk/core';
import type { OrchestrationMessageChunk } from './orchestration-message-chunk.js';
import type { ChatResult } from '@langchain/core/outputs';
import type { OrchestrationModuleConfig } from '@sap-ai-sdk/orchestration';
import type { BaseChatModelParams } from '@langchain/core/language_models/chat_models';
import type { ResourceGroupConfig } from '@sap-ai-sdk/ai-api';
import type { BaseMessage } from '@langchain/core/messages';
import type { CallbackManagerForLLMRun } from '@langchain/core/callbacks/manager';
import type { OrchestrationCallOptions } from './types.js';
import type { HttpDestinationOrFetchOptions } from '@sap-cloud-sdk/connectivity';

// TODO: Update all docs

/**
 * LangChain chat client for Azure OpenAI consumption on SAP BTP.
 */
export class OrchestrationClient extends BaseChatModel<
  OrchestrationCallOptions,
  OrchestrationMessageChunk
> {
  constructor(
    // Omit streaming until supported
    public orchestrationConfig: Omit<OrchestrationModuleConfig, 'streaming'>,
    public langchainOptions: BaseChatModelParams = {},
    public deploymentConfig?: ResourceGroupConfig,
    public destination?: HttpDestinationOrFetchOptions
  ) {
    super(langchainOptions);
  }

  _llmType(): string {
    return 'orchestration';
  }

  /**
   * Decisions:
   * bind only supports ParsedCallOptions, we don't support arbitrary LLM options, only tool calls & default BaseLanguageModelCallOptions, e.g. stop ✅
   * this aligns with other vendors' client designs (e.g. openai, google) ✅
   * top of the array (array[array.length - 1]) contains the current message, everything before then is history.
   * Module results are part of our own message type, which extends AI Message to work with all other langchain functionality.
   *
   * For timeout, we need to apply our own middleware, it is not handled by langchain. ✅.
   */

  override async _generate(
    messages: BaseMessage[],
    options: typeof this.ParsedCallOptions,
    runManager?: CallbackManagerForLLMRun
  ): Promise<ChatResult> {
    let caller = this.caller;
    if (options.maxConcurrency) {
      const { maxConcurrency, maxRetries, onFailedAttempt } =
        this.langchainOptions;
      caller = new AsyncCaller({
        maxConcurrency: maxConcurrency ?? options.maxConcurrency,
        maxRetries,
        onFailedAttempt
      });
    }
    const res = await caller.callWithOptions(
      {
        signal: options.signal
      },
      () => {
        // consider this.tools & this.stop property, merge it with template orchestration config
        const { inputParams } = options;
        const mergedOrchestrationConfig =
          this.mergeOrchestrationConfig(options);
        const orchestrationClient = new OrchestrationClientBase(
          mergedOrchestrationConfig,
          this.deploymentConfig,
          this.destination
        );
        const messagesHistory =
          mapLangchainMessagesToOrchestrationMessages(messages);
        const customRequestConfig: CustomRequestConfig = {
          ...options.customRequestConfig,
          middleware: resilience({ timeout: options.timeout })
        };
        return orchestrationClient.chatCompletion(
          {
            // how to handle tools here? doesn't really exist as input in orchestration as message history
            // make template a call option, to merge it ??
            messagesHistory,
            inputParams
          },
          customRequestConfig
        );
      }
    );

    const content = res.getContent();

    // we currently do not support streaming
    await runManager?.handleLLMNewToken(
      typeof content === 'string' ? content : ''
    );

    return mapOutputToChatResult(res.data);
  }

  private mergeOrchestrationConfig(
    options: typeof this.ParsedCallOptions
  ): OrchestrationModuleConfig {
    const { tools = [], stop = [] } = options;
    return {
      ...this.orchestrationConfig,
      llm: {
        ...this.orchestrationConfig.llm,
        model_params: {
          ...this.orchestrationConfig.llm.model_params,
          ...(stop.length && {
            stop: [
              ...(this.orchestrationConfig.llm.model_params?.stop || []),
              ...stop
            ]
          })
        }
      },
      templating: {
        ...this.orchestrationConfig.templating,
        ...(isTemplate(this.orchestrationConfig.templating) &&
          tools.length && {
            tools: [
              ...(this.orchestrationConfig.templating.tools || []),
              ...tools
            ]
          })
      }
    };
  }
}
