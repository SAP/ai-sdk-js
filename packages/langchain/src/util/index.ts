export * from './chunk-array.js';
