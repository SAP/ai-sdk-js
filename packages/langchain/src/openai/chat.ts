import { CallbackManagerForLLMRun } from '@langchain/core/callbacks/manager';
import { BaseMessage } from '@langchain/core/messages';
import type { ChatResult } from '@langchain/core/outputs';
import { AzureOpenAiChatClient as AzureOpenAiChatClientBase } from '@sap-ai-sdk/foundation-models';
import { BaseChatModel } from '@langchain/core/language_models/chat_models';
import { mapLangchainToAiClient, mapOutputToChatResult } from './util.js';
import type {
  AzureOpenAiChatCallOptions,
  AzureOpenAiChatModelParams
} from './types.js';

/**
 * LangChain chat client for Azure OpenAI consumption on SAP BTP.
 */
export class AzureOpenAiChatClient extends BaseChatModel<AzureOpenAiChatCallOptions> {
  temperature?: number | null;
  top_p?: number | null;
  logit_bias?: Record<string, any> | null | undefined;
  user?: string;
  presence_penalty?: number;
  frequency_penalty?: number;
  stop?: string | string[];
  max_tokens?: number;
  private openAiChatClient: AzureOpenAiChatClientBase;

  constructor(fields: AzureOpenAiChatModelParams) {
    super(fields);
    this.openAiChatClient = new AzureOpenAiChatClientBase(fields);
    this.temperature = fields.temperature;
    this.top_p = fields.top_p;
    this.logit_bias = fields.logit_bias;
    this.user = fields.user;
    this.stop = fields.stop;
    this.presence_penalty = fields.presence_penalty;
    this.frequency_penalty = fields.frequency_penalty;
    this.max_tokens = fields.max_tokens;
  }

  _llmType(): string {
    return 'azure_openai';
  }

  override async _generate(
    messages: BaseMessage[],
    options: typeof this.ParsedCallOptions,
    runManager?: CallbackManagerForLLMRun
  ): Promise<ChatResult> {
    const res = await this.caller.callWithOptions(
      {
        signal: options.signal
      },
      () =>
        this.openAiChatClient.run(
          mapLangchainToAiClient(this, messages, options),
          options.requestConfig
        )
    );

    const content = res.getContent();

    // we currently do not support streaming
    await runManager?.handleLLMNewToken(
      typeof content === 'string' ? content : ''
    );

    return mapOutputToChatResult(res.data);
  }
}
