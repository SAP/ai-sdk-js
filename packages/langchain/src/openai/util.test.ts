import {
  AIMessage,
  HumanMessage,
  RemoveMessage,
  SystemMessage,
  ToolMessage
} from '@langchain/core/messages';
import { parseMockResponse } from '../../../../test-util/mock-http.js';
import { mapLangchainToAiClient, mapOutputToChatResult } from './util.js';
import { AzureOpenAiChatClient } from './chat.js';
import type { BaseMessage } from '@langchain/core/messages';
import type {
  AzureOpenAiCreateChatCompletionResponse,
  AzureOpenAiCreateChatCompletionRequest
} from '@sap-ai-sdk/foundation-models';
import type { AzureOpenAiChatCallOptions } from './types.js';

// Signal and Prompt Index are provided by the super class in every call
const defaultOptions = {
  signal: undefined,
  promptIndex: 0
};

describe('Mapping Functions', () => {
  it('should parse an OpenAI response to a (LangChain) chat response', async () => {
    const openAiMockResponse =
      await parseMockResponse<AzureOpenAiCreateChatCompletionResponse>(
        'foundation-models',
        'azure-openai-chat-completion-success-response.json'
      );
    const result = mapOutputToChatResult(openAiMockResponse);
    expect(result).toMatchSnapshot();
  });

  it('should parse a LangChain input to an AI SDK input', async () => {
    const langchainPrompt: BaseMessage[] = [
      new ToolMessage('Tool Test Content', 'test-id'),
      new HumanMessage('Human Test Content'),
      new SystemMessage('System Test Content'),
      new AIMessage('AI Test Content')
    ];

    const request: AzureOpenAiCreateChatCompletionRequest = {
      messages: [
        {
          role: 'tool',
          content: 'Tool Test Content',
          tool_call_id: 'test-id'
        },
        {
          role: 'user',
          content: 'Human Test Content'
        },
        {
          role: 'system',
          content: 'System Test Content'
        },
        {
          role: 'assistant',
          content: 'AI Test Content'
        }
      ],
      tools: [{ type: 'function', function: { name: 'test', parameters: {} } }],
      tool_choice: 'auto',
      functions: [{ name: 'random' }, { name: 'test' }]
    };
    const client = new AzureOpenAiChatClient({ modelName: 'gpt-35-turbo' });
    const options: AzureOpenAiChatCallOptions = {
      tools: [{ type: 'function', function: { name: 'test', parameters: {} } }],
      tool_choice: 'auto',
      functions: [{ name: 'random' }, { name: 'test' }]
    };
    const mapping = mapLangchainToAiClient(client, langchainPrompt, {
      ...defaultOptions,
      ...options
    });
    expect(mapping).toMatchObject(request);
  });

  it('throws an error if the message type is not supported', async () => {
    const langchainPrompt: BaseMessage[] = [
      new RemoveMessage({ id: 'test-id' })
    ];
    const client = new AzureOpenAiChatClient({ modelName: 'gpt-35-turbo' });
    expect(() =>
      mapLangchainToAiClient(client, langchainPrompt, defaultOptions)
    ).toThrowErrorMatchingInlineSnapshot('"Unsupported message type: remove"');
  });
});
