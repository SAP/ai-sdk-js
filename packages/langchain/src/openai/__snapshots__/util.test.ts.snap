// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mapping Functions should parse an OpenAI response to a (LangChain) chat response 1`] = `
{
  "generations": [
    {
      "generationInfo": {
        "finish_reason": "stop",
        "function_call": undefined,
        "index": 0,
        "tool_calls": undefined,
      },
      "message": {
        "id": [
          "langchain_core",
          "messages",
          "AIMessage",
        ],
        "kwargs": {
          "additional_kwargs": {
            "finish_reason": "stop",
            "function_call": undefined,
            "index": 0,
            "tool_call_id": "",
            "tool_calls": undefined,
          },
          "content": "Hello! I'm just a computer program, so I don't have feelings, but thanks for asking. How can I assist you today?",
          "invalid_tool_calls": [],
          "response_metadata": {},
          "tool_calls": [],
        },
        "lc": 1,
        "type": "constructor",
      },
      "text": "Hello! I'm just a computer program, so I don't have feelings, but thanks for asking. How can I assist you today?",
    },
  ],
  "llmOutput": {
    "created": 1726839449,
    "id": "chatcmpl-A9Y77vG0OKsk9lUU0UAtlTKpnjyQ6",
    "model": "gpt-4o-2024-05-13",
    "object": "chat.completion",
    "tokenUsage": {
      "completionTokens": 26,
      "promptTokens": 13,
      "totalTokens": 39,
    },
  },
}
`;
